types/prefs/map.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"maps"
+	"net/netip"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"golang.org/x/exp/constraints"
+	"tailscale.com/types/opt"
+	"tailscale.com/types/ptr"
+	"tailscale.com/types/views"
+)
+
+// MapKeyType is a constraint allowing types that can be used as [Map] and [StructMap] keys.
+// To satisfy this requirement, a type must be comparable and must encode as a JSON string.
+// See [jsonv2.Marshal] for more details.
+type MapKeyType interface {
+	~string | constraints.Integer | netip.Addr | netip.Prefix | netip.AddrPort
+}
+
+// Map is a preference type that holds immutable key-value pairs.
+type Map[K MapKeyType, V ImmutableType] struct {
+	preference[map[K]V]
+}
+
+// MapOf returns a map configured with the specified value and [Options].
+func MapOf[K MapKeyType, V ImmutableType](v map[K]V, opts ...Options) Map[K, V] {
+	return Map[K, V]{preferenceOf(opt.ValueOf(v), opts...)}
+}
+
+// MapWithOpts returns an unconfigured [Map] with the specified [Options].
+func MapWithOpts[K MapKeyType, V ImmutableType](opts ...Options) Map[K, V] {
+	return Map[K, V]{preferenceOf(opt.Value[map[K]V]{}, opts...)}
+}
+
+// View returns a read-only view of m.
+func (m *Map[K, V]) View() MapView[K, V] {
+	return MapView[K, V]{m}
+}
+
+// Clone returns a copy of m that aliases no memory with m.
+func (m *Map[K, V]) Clone() *Map[K, V] {
+	res := ptr.To(*m)
+	if v, ok := m.s.Value.GetOk(); ok {
+		res.s.Value.Set(maps.Clone(v))
+	}
+	return res
+}
+
+// Equal reports whether m and m2 are equal.
+func (m Map[K, V]) Equal(m2 Map[K, V]) bool {
+	if m.s.Metadata != m2.s.Metadata {
+		return false
+	}
+	v1, ok1 := m.s.Value.GetOk()
+	v2, ok2 := m2.s.Value.GetOk()
+	if ok1 != ok2 {
+		return false
+	}
+	return !ok1 || maps.Equal(v1, v2)
+}
+
+// MapView is a read-only view of a [Map].
+type MapView[K MapKeyType, V ImmutableType] struct {
```

Comment:
is this (and its methods) generated with `tailscale.com/cmd/viewer`?

Reply:
No. It is a custom type similar to `tailscale.com/types/views.Map` and `tailscale.com/types/views.MapFn`, but specifically for preferences.
We updated `tailscale.com/cmd/viewer` to support generics, container types and views in a few preceding PRs: [[1]](https://github.com/tailscale/tailscale/pull/12738), [[2]](https://github.com/tailscale/tailscale/pull/12809), [[3]](https://github.com/tailscale/tailscale/pull/13147), etc.

Reply:
Gotcha. I know this may be disruptive for your plans, but could we layer the preference-specific logic (i.e. metadata) on top of the `types/views` package, instead of reimplementing a lot of these bits? With embedding, or just nesting types?

Reply:
We do layer the preference-specific logic on top of the `types/views` package. As a matter of fact, the view types in the `prefs` package have methods like `Value`, `ValueOk` and `DefaultValue`, which all return the corresponding view types from the `types/views` package. We do not re-implement any of that in this package.

But no, we could not use the `types/views` types in place of the custom view types we've got in the `prefs` package.

Reply:
Also: https://github.com/tailscale/tailscale/pull/12830#discussion_r1722595393